src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.cs

```diff
@@ -0,0 +1,127 @@
+            ClientCapabilities clientCapabilities, string? clientName, CancellationToken cancellationToken) where TRequestType : class
+        {
+            // Create a task completion source that will represent the processing of this request to the caller
+            var completion = new TaskCompletionSource<TResponseType>();
+
+            var item = new QueueItem(mutatesSolutionState, clientCapabilities, clientName,
+                callback: async (context) =>
+                {
+                    // Check if cancellation was requested while this was waiting in the queue
+                    if (cancellationToken.IsCancellationRequested)
+                    {
+                        completion.SetCanceled();
+                        return false;
+                    }
+
+                    try
+                    {
+                        var result = await handler.HandleRequestAsync(request, context, cancellationToken).ConfigureAwait(false);
+                        completion.SetResult(result);
+                        return true;
+                    }
+                    catch (OperationCanceledException)
+                    {
+                        completion.SetCanceled();
+                    }
+                    catch (Exception exception)
+                    {
+                        // Pass the exception to the task completion source, so the caller of the ExecuteAsync method can observe but
+                        // don't let it escape from this callback, so it doens't affect the queue processing.
+                        completion.SetException(exception);
+                    }
+                    return false;
+                });
+            _queue.Enqueue(item);
+
+            return completion.Task;
+        }
+
+        private async Task ProcessQueueAsync()
+        {
+            while (true)
+            {
+                var work = await _queue.DequeueAsync().ConfigureAwait(false);
+
+                var solution = GetCurrentSolution();
+                solution = MergeChanges(solution, _lastMutatedSolution);
+
+                Solution? mutatedSolution = null;
+                var context = new RequestContext(solution, s => mutatedSolution = s, work.ClientCapabilities, work.ClientName);
+
+                if (work.MutatesSolutionState)
+                {
+                    // Mutating requests block other requests from starting to ensure an up to date snapshot is used.
+                    var ranToCompletion = await work.Callback(context).ConfigureAwait(false);
+
+                    // If the handling of the request failed, the exception will bubble back up to the caller, but we
+                    // still need to react to it here by throwing away solution updates
+                    if (ranToCompletion)
+                    {
+                        _lastMutatedSolution = mutatedSolution ?? _lastMutatedSolution;
```

Comment:
There are really 3 options for what to do if a mutating request fails:
1. Throw away the mutations and hope for the best (which is what this code is doing)
2. Apply the mutation and hope for the best
3. Environment.FailFast

I don't have a strong feeling either way really, as I think 1 and 2 both have an equal chance of resulting in a solution state in memory that doesn't match what the user is seeing, but I went with 1 because I am working on the theory that people will be more likely to understand if the server seems to be "lagging" the state they're seeing, and try reconnecting or re-opening the document.

Note that the user experience for errors is unchanged in this PR, so they'll still see the yellow bar etc. and know that something went wrong.